data-prepper-plugins/key-value-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java

```diff
@@ -179,79 +240,215 @@ private void validateKeySets(final Set<String> includeSet, final Set<String> exc
     @Override
     public Collection<Record<Event>> doExecute(final Collection<Record<Event>> records) {
         for(final Record<Event> record : records) {
-            final Map<String, Object> parsedMap = new HashMap<>();
+            final Map<String, Object> outputMap = new HashMap<>();
             final Event recordEvent = record.getData();
-
             final String groupsRaw = recordEvent.get(keyValueProcessorConfig.getSource(), String.class);
             final String[] groups = fieldDelimiterPattern.split(groupsRaw, 0);
 
-            for(final String group : groups) {
-                final String[] terms = keyValueDelimiterPattern.split(group, 2);
-                String key = terms[0];
-                Object value;
-
-                if (!includeKeysSet.isEmpty() && !includeKeysSet.contains(key)) {
-                    LOG.debug("Skipping not included key: '{}'", key);
-                    continue;
+            if (keyValueProcessorConfig.getRecursive()) {
+                ObjectMapper mapper = new ObjectMapper();
```

Comment:
You do not need to create a new Mapper per record. Refactor it into an instance attribute:

`private final ObjectMapper mapper = ...`